Cython/Compiler/TypeSlots.py

```diff
@@ -503,6 +503,23 @@ def spec_value(self, scope):
         return self.slot_code(scope)
 
 
+class DelSlot(SyntheticSlot):
+    # For __del__ if defined
+    def __init__(self, slot_name, user_methods, default_value, **kwargs):
+        super(DelSlot, self).__init__(slot_name, user_methods, default_value, **kwargs)
+
+    def slot_code(self, scope):
+        if not scope.lookup_here("__del__"):
+            return 0
+        return """
+        #if CYTHON_USE_TP_FINALIZE
+        %s
+        #else
+        NULL
+        #endif
+        """ % InternalMethodSlot.slot_code(self, scope)
```

Comment:
I'd rather not have this part right here, but in a new option similar to `ifdef=`. `slot_code()` should return the plain value, `generate()` takes care of the C-level conditions.

Reply:
@ax487 As a hint (since I think it was me that originally suggested putting this `#if` mess in `slot_code`): https://github.com/cython/cython/blob/f4ce85ba8de0fc26a1e47d1bb6e0027ec95fe24a/Cython/Compiler/TypeSlots.py#L294

Reply:
I've tried to solve this in https://github.com/cython/cython/commit/f7570d7b7bc8b5cac0ae7f567b99a9e3c5826521 (in the interests of getting this change finished).

I don't think I can push it to this PR myself so @scoder could you apply it (if it looks suitable)?